### PR TITLE
net: lwm2m: Add a data validation callback

### LINKED ISSUE #3931: LWM2M: Data Validation Callback
**_Reported by Michael Scott:_**

Smart Objects need a way to validate the incoming data prior to having data element values being changed.

Create a new callback which can return a 0 or non-zero value to indicate that the new data is acceptable for the engine to assign.

(Imported from Jira ZEP-2500)

### PR DESCRIPTION
**Pull request by rlubos**
Add a data validation callback to the resource structure, which can be
registered by an application. It allows to verify the data before
actually modifying the resource data.

If the callback is registered for a resource, the data is decoded into a
temporary buffer first, and only copied into the actual resource buffer
if the validation is successful. If no validation is required (and thus
no callback registered) the resource value is decoded directly into the
resource buffer, as it used to be.

Resolves #3931

### CODE REVIEW

**Review by jukkar — APPROVED** [review]

**Comment by pfalcon** [comment]
> If the callback is registered for a resource, the data is decoded into a
temporary buffer first, and only copied into the actual resource buffer
if the validation is successful.

This seems like a pretty heavy-weight approach. Were other approaches considered? E.g., when a validation func get "undecoded" data? I understand that doing this way pushes users to break abstraction layers (or do double-work of decoding), but at this time, it's unclear how demanded this feature is at all.

Assuming it should be done line that, was it considered to make CONFIG_LWM2M_ENGINE_VALIDATION_BUFFER_SIZE == 0 to mean "no validation support" (and thus `#if CONFIG_LWM2M_ENGINE_VALIDATION_BUFFER_SIZE == 0` sprinkled around to disable unneeded code in this case)?

File: include/net/lwm2m.h [hunk 1]
```diff
@@ -746,6 +753,29 @@ int lwm2m_engine_register_read_callback(char *pathstr,
...
+ *
+ * @note All resources that have a validation callback registered are initially
+ *       decoded into a temporary validation buffer. Make sure that
+ *       ``CONFIG_LWM2M_ENGINE_VALIDATION_BUFFER_SIZE`` is large enough to
+ *       store each of the validated resources (individually).
+ *
+ * @param[in] pathstr LwM2M path string "obj/obj-inst/res(/res-inst)"
+ * @param[in] cb Validate resource data callback
+ *
+ * @return 0 if the data is accepted or a negative error code in case data is
```
Comment (pfalcon) [inline include/net/lwm2m.h:773]: This doesn't look like describing return value of lwm2m_engine_register_validate_callback() (like should be done here), but rather of the validation callback.

File: subsys/net/lib/lwm2m/Kconfig [hunk 2]
```diff
@@ -56,6 +56,16 @@ config LWM2M_ENGINE_MESSAGE_HEADER_SIZE
 	help
 	  Extra room allocated to handle CoAP header data
 
+config LWM2M_ENGINE_VALIDATION_BUFFER_SIZE
```
Comment (pfalcon) [inline subsys/net/lib/lwm2m/Kconfig:59]: As mentioned in the top-level comment, suggestion is to use CONFIG_LWM2M_ENGINE_VALIDATION_BUFFER_SIZE == 0 as "validation support is disabled".

File: subsys/net/lib/lwm2m/lwm2m_engine.c [hunk 3]
```diff
@@ -2350,30 +2374,59 @@ size_t lwm2m_engine_get_opaque_more(struct lwm2m_input_context *in,
...
 		}
+
+		if (res->validate_cb) {
+			ret = res->validate_cb(
+				obj_inst->obj_inst_id, res->res_id,
+				res_inst->res_inst_id, write_buf, len,
+				last_pkt_block && last_block, opaque_ctx.len);
+			if (ret < 0) {
+				/* Bad Request. */
+				return -EEXIST;
```
Comment (pfalcon) [inline subsys/net/lib/lwm2m/lwm2m_engine.c:2424]: Why such an error code, if above -EINVAL is used?

**Review by pfalcon — COMMENTED** [review]
Some generic comments. Generally, I'd be happy to give +1, just some general concern of feature creep and constant growth of the code/data size. E.g., currently there were:

> lwm2m_engine_get_data_cb_t		pre_write_cb;
	lwm2m_engine_set_data_cb_t		validate_cb;
	lwm2m_engine_set_data_cb_t		post_write_cb;

Couldn't one of the pre_write_cb/post_write_cb take care of task of validation? And assuming that existing Zephyr lwm2m usages did take care of validation using them, why burden them with additional code growth for features they already have covered? Hence comment about making this support optional (but looking thru the entire patchset, I agree it would be boring activity to add all those #if's around).

So again, just generic comments for your consideration.

File: include/net/lwm2m.h (hunk 1 above)
Comment (rlubos) [reply include/net/lwm2m.h:773]: Indeed, will fix.

File: subsys/net/lib/lwm2m/Kconfig (hunk 2 above)
Comment (rlubos) [reply subsys/net/lib/lwm2m/Kconfig:59]: Makes sense, I'll implment it like this.

File: subsys/net/lib/lwm2m/lwm2m_engine.c (hunk 3 above)
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_engine.c:2424]: `lwm2m_engine_set()` where I return -EINVAL is a function that can be invoked from the application to set a resource data, hence it makes sense to return -EINVAL back to the application which suggests the data provided was not valid.

Here, I validate the data that comes from the network, and the engine generates the response based on the error code here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/net/lib/lwm2m/lwm2m_engine.c#L4000
Not very intuitive, but in order to gerate BAD REQUEST response, `-EEXIST` error code is expected.

**Comment by rlubos** [comment]
>This seems like a pretty heavy-weight approach. Were other approaches considered? E.g., when a validation func get "undecoded" data?

I considered such an approach initially, to validate the data directly from the message buffer, but different encodings used in the LwM2M turned out to be a blocker here. For instance, TLV and JSON encodings will encode integers differently (TLV encodes them in binary while JSON encodes them as a string) so the actual validation callback would need to get the actual information regarding the encoding used, plus would need to handle different encodings. IMO that's too much hassle for the application developers, so I've decided to introduce the intermediary buffer for decoding.

Note that this buffer will only be used if the validation callback is registered for a resource. If it's not, we decode the data directly into the resurce data buffer as it used to be done.

> I understand that doing this way pushes users to break abstraction layers (or do double-work of decoding), but at this time, it's unclear how demanded this feature is at all. 

I guess it's not the most demanded feature as the issue reuquesting it was pretty old and there were not much interest. Still, those issues (there are still quite a few) need either to be implemented or closed and I found this one a valuable improvement.

**Comment by pfalcon** [comment]
Thanks for the responses.

> Note that this buffer will only be used if the validation callback is registered for a resource.

Yes, but buffer is reserved nonetheless, and it defaults to 64 bytes, not exactly small size.

> Still, those issues (there are still quite a few) need either to be implemented or closed and I found this one a valuable improvement.

Right, and I from my side just trying to do "decent review", not add more workload to your queue ;-). So, please use your maintainer's judgement regarding the suggestions, if they're too much effort to implement for now and/or pose maintenance complications, then well, it's ok as it is.

**Comment by rlubos** [comment]
@pfalcon I've addressed your comments.

> Couldn't one of the pre_write_cb/post_write_cb take care of task of validation? And assuming that existing Zephyr lwm2m usages did take care of validation using them, why burden them with additional code growth for features they already have covered? 

The validation callback fill the gap between the other two callbacks. The `pre_write_cb` is only used to obtain a data buffer (before decoding the data) and `post_write_cb` is called after the data is decoded into the resource buffer. Theoretically, one could use `post_write_cb` to validate the data, but it has it isn't the best approach as at this point, if the new data is considered invalid, the old resoruce value is already lost. 

Anyway, I've done some ifdeffing in the code to make the validation callback optional, based on the `lwm2m_client` sample it allows to save a few hundred bytes of RAM and FLASH if disabled (though the actual number may differ based on the amount of resources in the system).

**Review by pfalcon — APPROVED** [review]
Thanks!

**Pull request merged**